player/lua/input.lua

```diff
@@ -0,0 +1,77 @@
+--[[
+This file is part of mpv.
+
+mpv is free software; you can redistribute it and/or
+modify it under the terms of the GNU Lesser General Public
+License as published by the Free Software Foundation; either
+version 2.1 of the License, or (at your option) any later version.
+
+mpv is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+]]
+
+local utils = require "mp.utils"
+local input = {}
+
+function input.get(t)
+    local console_enabled = mp.get_property_native("user-data/console/enabled")
+    if console_enabled == nil then
+        return nil, "The console script is not loaded."
+    end
+    if console_enabled then
+        return nil, "The console is already open."
+    end
+
+    local args = utils.format_json({
+        prompt = t.prompt,
+        default_text = t.default_text,
+        cursor_position = t.cursor_position,
+        id = t.id,
+    })
+
+    mp.commandv("script-message-to", "console", "get-input",
+                mp.get_script_name(), args)
+
+    mp.register_script_message("input-event", function (type, text, cursor_position)
+        if t[type] then
+            local suggestions, completion_start_position = t[type](text, cursor_position)
```

Comment:
`cursor_position` is supposed to be a number right? But when the number is sent to `script-message` it gets converted into a string, so `cursor_position` will be a string that contains a number. This may be confusing, so it may be worth converting the string back to a number.

Also if you wanted to future proof this code against any other arguments that might be sent by `closed` or any other callback and maintain the correct types you could switch to using a JSON array:

```suggestion
    mp.register_script_message("input-event", function (type, args)
        if t[type] then
            local suggestions, completion_start_position = t[type](unpack( (utils.parse_json(args)) ))
```

Then in `console.lua` you just do:
```lua
    mp.commandv('script-message-to', input_caller, 'input-event',
                'closed', utils.format_json({line, cursor}))
```

Reply:
Added `tonumber`. I initiallity sanitized the cursor position like `show_and_type` actually, but then decided I didn't want to make the PR even bigger.

I don't think we need to switch to JSON prematurely when we may never have many arguments. That is a private API that we can change whenever we want anyway.